### PR TITLE
[WIP] Use commonmarker instead of redcarpet

### LINKED ISSUE #2768: Refactor markdown handling to use commonmark
Markdown is not specified unambiguously, and we've had a lot of bugs and issues related to things that are a bit fuzzy, and where redcarpet handles it in one way, but people are used to it being handled differently on GitHub.

The markdown stuff has grown a bit over time, and we might also be able to simplify it.

It might be worth doing a rewrite using commonmark instead of redcarpet. See discussion in https://github.com/exercism/exercism.io/issues/2759

### PR DESCRIPTION
**Pull request by abhishekbose87**
fix #2768

### CODE REVIEW

**Comment by abhishekbose87** [comment]
@kytrinyx @mhelmetag I have currently made changes in tests related to sanitisation and extra newline characters. Please do let me know if my assumptions are wrong.
Also would like your feedback regarding the refactoring changes I have made.

**Comment by mhelmetag** [comment]
@abhishekbose87 , looks pretty good to me but it looks like travis failed because commonmarker wants a higher version of cmake than our build container currently has. I wonder if updating to 14.04 or something would help that (looks like it's 12.04 right now). Is that even possible? To specific ubuntu builds for travis? Idk...

https://github.com/travis-ci/travis-ci/issues/2030

**Comment by kytrinyx** [comment]
I'm trying the cmake upgrade thing in a branch, I'll report back here how it goes.

**Comment by kytrinyx** [comment]
The cmake upgrade hasn't worked out: https://travis-ci.org/exercism/exercism.io/builds/114065807

**Comment by mhelmetag** [comment]
>E: Could not open lock file /var/lib/dpkg/lock - open (13: Permission denied)
E: Unable to lock the administration directory (/var/lib/dpkg/), are you root?

Needs sudo

**Comment by kytrinyx** [comment]
It's using sudo. This is the .travis.yml file:

```
language: ruby
sudo: required
before_install:
    - sudo apt-get update && apt-get install --only-upgrade cmake
rvm:
  - 2.2.2
script:
  - 'createuser exercism --superuser && createdb -O exercism exercism_test'
  - 'RACK_ENV=test CI=1 bundle exec rake db:migrate test_with_coverage'
  - '( cd frontend && npm install && lineman spec-ci )'
  - "! git grep ' $' -- \\*.rb"
cache:

```

**Comment by mhelmetag** [comment]
@abhishekbose87 , you should be able to `git fetch upstream && git rebase upstream master` now so that `commonmarker` installs correctly and your build can run.

**Comment by abhishekbose87** [comment]
@mhelmetag That worked. Thanks a ton :+1: 
@kytrinyx Can you please review once if any changes are required.

**Comment by mhelmetag** [comment]
Glad it worked! Excited to finally see this markdown thing solved :smile_cat:

**Comment by kytrinyx** [comment]
Yepp, I'll review this today.

**Comment by mhelmetag** [comment]
@abhishekbose87 , just to give you an update, the heroku buildpack needs to be tweaked a bit (much like travis did) to upgrade cmake in order to install commonmarker. I was going to work on this sometime this week.

**Comment by kytrinyx** [comment]
@mhelmetag Thanks for checking back here with the update--I completely forgot to do that.

**Comment by k4rtik** [comment]
This should take care of #2759 as well. cf. https://github.com/exercism/xcpp/pull/72

**Comment by kytrinyx** [comment]
Yeah, 2759 is the same issue.

**Comment by abhishekbose87** [comment]
@mhelmetag @kytrinyx Sorry that I did not get time to follow up on this one. 

Is there some work left before this gets reviewed ?

**Comment by mhelmetag** [comment]
So... this is because commonmarker uses a version of cmake that doesn't come with the stock ruby heroku buildpack.

I couldn't get exercism properly deployed to my own heroku instance to play around with cmake buildpacks...

I opened this issue https://github.com/vmg/redcarpet/issues/546 awhile ago (our current markdown preprocessor) and got no response from the maintainer... but another person found that setting `lax_spacing: true` does help.

So I went ahead and tested that now:
Unprocessed:
screen shot 2016-06-06 at 9 20 06 pm

Processed:
screen shot 2016-06-06 at 9 20 11 pm

Looks pretty good now.

Basically, I'm just a little hesitant in using some random persons cmake buildpack for heroku when we can make this small change. But if there are other issues with redcarpet than just this... maybe it's worth doing. I leave the options up to @kytrinyx.

**Comment by kytrinyx** [comment]
> I'm just a little hesitant in using some random persons cmake buildpack for heroku when we can make this small change.

Agreed, especially if they're not responding to issues.

I'd say if we can do it with the small change, let's do it. We'll eventually find more problems if there are any, and can re-evaluate at that time.

**Comment by mhelmetag** [comment]
And this PR can be closed because of #2943

**Comment by kytrinyx** [comment]
Thanks @mhelmetag!
